transport/packet/src/v1.rs

```diff
@@ -0,0 +1,263 @@
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_u64(self.0)
+    }
+}
+
+impl<'a> serde::Deserialize<'a> for Tag {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'a>,
+    {
+        let value = u64::deserialize(deserializer)?;
+        Ok(Self(value))
+    }
+}
+
+impl Add for Tag {
+    type Output = Self;
+
+    fn add(self, other: Self) -> Self::Output {
+        Self(self.0 + other.0)
+    }
+}
+
+impl Sub for Tag {
+    type Output = Self;
+
+    fn sub(self, other: Self) -> Self::Output {
+        Self(self.0 - other.0)
+    }
+}
+
+impl std::fmt::Display for Tag {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(f, "Tag({})", self.0)
+    }
+}
+
+impl From<u64> for Tag {
+    fn from(value: u64) -> Self {
+        Self(value)
+    }
+}
+
+impl From<u32> for Tag {
+    fn from(value: u32) -> Self {
+        Self(value as u64)
+    }
+}
+
+impl From<u8> for Tag {
+    fn from(value: u8) -> Self {
+        Self(value as u64)
+    }
+}
+
+impl PartialEq<u64> for Tag {
+    fn eq(&self, other: &u64) -> bool {
+        self.0 == *other
+    }
+}
+
+impl PartialOrd<u64> for Tag {
+    fn partial_cmp(&self, other: &u64) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(other)
+    }
+}
+
+/// Resolved tag type with translated tag annotation.
+#[repr(u64)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum ResolvedTag {
+    Reserved(ReservedTag),
+    Custom(CustomTag),
+}
+
+impl From<ResolvedTag> for Tag {
+    fn from(tag: ResolvedTag) -> Self {
+        match tag {
+            ResolvedTag::Reserved(reserved_tag) => reserved_tag.into(),
+            ResolvedTag::Custom(custom_tag) => custom_tag.0,
+        }
+    }
+}
+
+impl From<Tag> for ResolvedTag {
+    fn from(tag: Tag) -> Self {
+        if tag < ReservedTag::UPPER_BOUND {
+            match tag.0 {
+                x if x == ReservedTag::SessionInit as u64 => ResolvedTag::Reserved(ReservedTag::SessionInit),
+                x if x == ReservedTag::Ping as u64 => ResolvedTag::Reserved(ReservedTag::Ping),
+                _ => ResolvedTag::Reserved(ReservedTag::Undefined),
+            }
+        } else {
+            ResolvedTag::Custom(CustomTag(tag))
+        }
+    }
+}
+
+/// List of all reserved application tags for the protocol.
+#[repr(u64)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+
+pub enum ReservedTag {
+    /// Opening a new session.
+    SessionInit = 0,
+    /// Ping traffic for 0-hop detection.
+    Ping = 14,
+    /// Undefined catch all.
+    Undefined = Self::UPPER_BOUND.0 - 1,
+}
+
+impl ReservedTag {
+    /// The upper limit value for the session reserved tag range.
+    pub const UPPER_BOUND: Tag = Tag(1 << 4); // 16
+}
+
+impl From<ReservedTag> for Tag {
+    fn from(tag: ReservedTag) -> Self {
+        Self(tag as u64)
+    }
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct CustomTag(Tag);
```

Comment:
Docs missing.

Reply:
@coderabbitai  Please create an issue to go through these tags and unify them in terms of a single `enum` based structure. Assign the feature to me, please.